### PR TITLE
Separate metadata fetch from `ArrowReaderBuilder` construction (#4674)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Related to https://github.com/apache/arrow-rs/discussions/4674

# Rationale for this change
 

This makes it easier to load the metadata once, and then use it to construct multiple readers

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 1]
```diff
@@ -234,48 +221,187 @@ impl ArrowReaderOptions {
...
-pub struct SyncReader<T: ChunkReader>(SerializedFileReader<T>);
+pub struct SyncReader<T: ChunkReader>(T);
 
 /// A synchronous builder used to construct [`ParquetRecordBatchReader`] for a file
 ///
 /// For an async API see [`crate::arrow::async_reader::ParquetRecordBatchStreamBuilder`]
 pub type ParquetRecordBatchReaderBuilder<T> = ArrowReaderBuilder<SyncReader<T>>;
 
-impl<T: ChunkReader + 'static> ArrowReaderBuilder<SyncReader<T>> {
+impl<T: ChunkReader + 'static> ParquetRecordBatchReaderBuilder<T> {
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:285]: This does not change the type, but improves the docs rendering as the methods will be shown for the typedef

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 2]
```diff
@@ -234,48 +221,187 @@ impl ArrowReaderOptions {
...
+    /// Returns the arrow [`SchemaRef`] for this parquet file
+    pub fn schema(&self) -> &SchemaRef {
+        &self.schema
+    }
+}
+
 #[doc(hidden)]
 /// A newtype used within [`ReaderOptionsBuilder`] to distinguish sync readers from async
-pub struct SyncReader<T: ChunkReader>(SerializedFileReader<T>);
+pub struct SyncReader<T: ChunkReader>(T);
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:278]: Continuing to use SerializedFileReader would have meant adding APIs to load metadata and then construct it from said metadata. It seemed simpler to just use the lower level SerializedPageReader as is done by the async API. This does mean the arrow readers no longer make use of the file APIs, but I think this is fine

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 3]
```diff
@@ -234,48 +221,187 @@ impl ArrowReaderOptions {
...
-            false => SerializedFileReader::new(reader)?,
-        };
+        let metadata = Self::load_metadata(&reader, options)?;
+        Ok(Self::new_with_metadata(reader, metadata))
+    }
+
+    /// Loads [`ArrowReaderMetadata`] from the provided [`ChunkReader`]
+    ///
+    /// See [`Self::new_with_metadata`] for how this can be used
+    pub fn load_metadata(
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:327]: I'm not totally sold on this method naming, nor whether it would be better to define a `ArrowReaderMetadata::load` and `ArrowReaderMetadata::load_async`.

**Comment by RinChanNOWWW** [comment]
LGTM. Thanks!

**Comment by tustvold** [comment]
Why the API change label? I don't believe this breaks any public APIs?

File: parquet/src/file/metadata.rs [hunk 4]
```diff
@@ -155,13 +155,13 @@ impl ParquetMetaData {
     }
 
     /// Override the column index
-    #[allow(dead_code)]
+    #[cfg(feature = "arrow")]
```
Comment (alamb) [inline parquet/src/file/metadata.rs:158]: 👍

File: parquet/src/file/serialized_reader.rs [hunk 5]
```diff
@@ -242,11 +242,6 @@ impl<R: 'static + ChunkReader> SerializedFileReader<R> {
             })
         }
     }
-
-    #[cfg(feature = "arrow")]
```
Comment (alamb) [inline parquet/src/file/serialized_reader.rs:246]: what is the purpose of this change? Perhaps we can mark it as deprecated for a while before removing it?

File: parquet/src/arrow/arrow_reader/mod.rs (hunk 3 above)
Comment (alamb) [reply parquet/src/arrow/arrow_reader/mod.rs:327]: I personally think it would be easier to reason about if the code to load metadata was on the `ArrowReaderMetadata` struct rather than on the related (but different) `ParquetRecordBatchReaderBuilder` etc

Thus I like the idea of `ArrowReaderMetadata::load` and `ArrowReaderMetadata::load_async` but I don't feel super strongly about this -- maybe others do

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 6]
```diff
@@ -234,48 +221,187 @@ impl ArrowReaderOptions {
     }
 }
 
+/// The clone-able metadata necessary to construct a [`ArrowReaderBuilder`]
```
Comment (alamb) [inline parquet/src/arrow/arrow_reader/mod.rs:224]:
```suggestion
/// The metadata necessary to construct a [`ArrowReaderBuilder`]. 
///
/// This structure is inexpensive to clone. 
```

**Review by alamb — APPROVED** [review]
What is the context for this PR? I feel like there must be some backstory but it doesn't appear to be linked. Update: I found the related conversation https://github.com/apache/arrow-rs/discussions/4674 and linked it here)


I think the code looks good and well documented to me. 

I marked this PR as an `api-change` as I think it is -- let me know if you disagree 

Finally, I think this PR would be strongly  with  a test / example of the stated usecase: read metadata once, but create multiple readers for the same file.

cc @Dandandan and @thinkharderdev  who I think are knowledgable about the use of this functionality upstream in DataFusion

**Comment by tustvold** [comment]
> I marked this PR as an api-change as I think it is -- let me know if you disagree

Where is the breaking change you are refering to?

> Finally, I think this PR would be strongly with a test / example of the stated usecase: read metadata once, but create multiple readers for the same file.

https://github.com/apache/arrow-rs/pull/4676/files#diff-850b3a44587149637b8545f66603a2b1252959fd36f7ddc55f37d6b5357816c6R374

?

File: parquet/src/file/serialized_reader.rs (hunk 5 above)
Comment (alamb) [reply parquet/src/file/serialized_reader.rs:246]: ah, I missed that it was not `pub` 👍

**Comment by alamb** [comment]
> >  I marked this PR as an api-change as I think it is -- let me know if you disagree

> Where is the breaking change you are refering to?

I was thinking of https://github.com/apache/arrow-rs/pull/4676/files#r1289997977 but I suppose that since `SyncReader` isn't documented it is unlikely that anyone is using it directly 🤔  


> >  Finally, I think this PR would be strongly with a test / example of the stated usecase: read metadata once, but create multiple readers for the same file.

> https://github.com/apache/arrow-rs/pull/4676/files#diff-850b3a44587149637b8545f66603a2b1252959fd36f7ddc55f37d6b5357816c6R374

As sorry I missed that.

**Comment by tustvold** [comment]
>  was thinking of https://github.com/apache/arrow-rs/pull/4676/files#r1289997977 but I suppose that since SyncReader isn't documented it is unlikely that anyone is using it directly thinking

The tuple field is not public, so cannot be used by external code

**Review by thinkharderdev — APPROVED** [review]
Makes sense to me

File: parquet/src/arrow/arrow_reader/mod.rs (hunk 3 above)
Comment (tustvold) [reply parquet/src/arrow/arrow_reader/mod.rs:327]: > to me if it was on ParquetRecordBatchReaderBuilder

Do you mean ArrowReaderMetadata?

Comment (thinkharderdev) [reply parquet/src/arrow/arrow_reader/mod.rs:327]: Yeah, sorry that's what I meant.
